### PR TITLE
swapped-to-immutable-record-for-dto

### PR DESCRIPTION
**Pull request by anetteOlli**
My idea of fun on a friday night is reading up on java 17 new record class for DTOs. 

It was a fun exercise swapping the OrcidCredential class with orcidCredential record. So I   wanted to commit my changes and share it with you.

### CODE REVIEW

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java [hunk 1]
```diff
@@ -1,118 +1,43 @@
...
-
-    private static final String ORCID = "orcid";
-    private static final String ACCESS_TOKEN = "accessToken";
-    private static final String TOKEN_TYPE = "tokenType";
-    private static final String EXPIRES_IN = "expiresIn";
-    private static final String TOKEN_VERSION = "tokenVersion";
-    private static final String PERSISTENT = "persistent";
-    private static final String ID_TOKEN = "idToken";
-    private static final String TOKEN_ID = "tokenId";
-    private static final String MODIFIED = "modified";
```
Comment (anetteOlli) [inline orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:33]: All this scaffolding: GONE!

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java [hunk 2]
```diff
@@ -1,118 +1,43 @@
...
-    private Instant modified;
-
-    @JsonProperty(CREATED)
-    private Instant created;
-
-    @JacocoGenerated
-    public OrcidCredentials() {
-    }
-
-    public OrcidCredentials copy() {
```
Comment (anetteOlli) [inline orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:70]: The copy method is no longer needed. We can't modify a record in any way, so no need for deep copy before we modify.

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java [hunk 3]
```diff
@@ -1,118 +1,43 @@
...
-    buildMethodName = "build",
-    setterPrefix = "with"
+        builderClassName = "OrcidCredentialsBuilder",
+        toBuilder = true,
+        builderMethodName = "builder",
+        buildMethodName = "build",
+        setterPrefix = "with"
 )
-@AllArgsConstructor(access = AccessLevel.PACKAGE)
-public class OrcidCredentials implements JsonSerializable {
```
Comment (anetteOlli) [inline orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:23]: Notice the record can still implement interfaces.
(you can't extend tho)

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/service/OrcidServiceImpl.java [hunk 4]
```diff
@@ -28,16 +32,30 @@ public OrcidServiceImpl(String orcidTableName, AmazonDynamoDB client, Clock cloc
...
+
+        var orcidCredentialsWithTimeStamp = createOrcidCredentialsWithTimeStamp(orcidCredentials);
+        return insertOrcidCredentials(orcidCredentialsWithTimeStamp);
+    }
+
+    private OrcidCredentials createOrcidCredentialsWithTimeStamp(OrcidCredentials orcidCredentials) {
         var currentTime = clockForTimestamps.instant();
-        orcidCredentials.setModified(currentTime);
-        orcidCredentials.setCreated(currentTime);
-        return insertOrcidCredentials(orcidCredentials);
```
Comment (anetteOlli) [inline orcid-commons/src/main/java/no/sikt/nva/orcid/commons/service/OrcidServiceImpl.java:34]: Not setters in a record! 

(You can make a weird workaround inside the record, but anyone who does that has lost the point of records IMO)

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java [hunk 5]
```diff
@@ -1,118 +1,43 @@
...
+public record OrcidCredentials(URI orcid,
+                               String accessToken,
+                               String tokenType,
+                               int expiresIn,
+                               String tokenVersion,
+                               boolean persistent,
+                               String idToken,
+                               int tokenId,
+                               Instant modified,
+                               Instant created) implements JsonSerializable {
```
Comment (brinxmat) [inline orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:31]: Since we're on the topic, we have 10 fields in this record, which leads me to believe that we're missing something in the design…e.g. we have a boolean "persistent" and a field "expiresIn"…are these related?

**Review by brinxmat — APPROVED** [review]

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java (hunk 5 above)
Comment (brinxmat) [reply orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:31]: The PR clearly explains the benefits of records, great work. We should have a quick run through of records + serialisation of these on Friday, please :) 

I appreciate the PR, btw. It's unfortunate that we have to rely on personal interest and Friday evenings to make these BIG improvements to the code, but this definitely brings the benefits of moving to Java 17 front-and-centre. Thank you :) make sure you submit the hours on your timesheet.

**Review by torarnet — APPROVED** [review]

File: orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java (hunk 5 above)
Comment (anetteOlli) [reply orcid-commons/src/main/java/no/sikt/nva/orcid/commons/model/business/OrcidCredentials.java:31]: Regarding the properties names: yes they are terrible. 

This is how they are served from Orcid (so yes there is a tokenId and an IdToken best namin ever!), and the documentation of what we are supposed to store doesn't match any of the the properties names we  received. So version 1: just store them as is.

**Review by StigNorland — APPROVED** [review]

**Review by ketilaa — APPROVED** [review]
Looks great ;)
